org.knime.scijava.commands/src/org/knime/scijava/commands/dialog/DefaultDialogService.java

```diff
@@ -0,0 +1,91 @@
+package org.knime.scijava.commands.dialog;
+
+import java.util.List;
+
+import javax.swing.JPanel;
+
+import org.knime.core.util.Pair;
+import org.knime.scijava.commands.MultiOutputListener;
+import org.knime.scijava.commands.widget.SettingsModelWidgetModel;
+import org.scijava.module.Module;
+import org.scijava.module.ModuleException;
+import org.scijava.module.ModuleInfo;
+import org.scijava.module.ModuleItem;
+import org.scijava.module.ModuleService;
+import org.scijava.plugin.Parameter;
+import org.scijava.plugin.Plugin;
+import org.scijava.service.AbstractService;
+import org.scijava.ui.swing.widget.SwingInputHarvester;
+import org.scijava.widget.InputPanel;
+import org.scijava.widget.InputWidget;
+import org.scijava.widget.WidgetModel;
+import org.scijava.widget.WidgetService;
+
+/**
+ * Default implementation of {@link DialogService}.
+ *
+ * @author Christian Dietz, University of Konstanz
+ */
+@Plugin(type = DialogService.class)
+public class DefaultDialogService extends AbstractService
+        implements DialogService {
+
+    @Parameter
+    private WidgetService ws;
+
+    @Parameter
+    private ModuleService ms;
+
+    @Override
+    public Pair<JPanel, List<SettingsModelWidgetModel>> dialogPanel(
+            final Module module) {
+
+        final SwingInputHarvester builder = new SwingInputHarvester() {
+            @SuppressWarnings("unchecked")
+            @Override
+            public void buildPanel(final InputPanel<JPanel, JPanel> inputPanel,
+                    final Module module) throws ModuleException {
+
+                // automatically resolve special fields
+                // FIXME can be extended using plugin mechanism?
```

Comment:
@dietzc Do you mean adding a plugin type for this? In Pure-Scijava, resolving module items before the UI gets to them  is the Preprocessors job, which already is a plugin, but not useable for our use case.